lib/script.js

```diff
@@ -40,106 +37,111 @@ Script.prototype.run = function(methodName, globals) {
 };
 
 Script.prototype.createTimeout = function(methodName) {
-  var _this = this;
-
-  if (this.timeoutId){
+  if (this.timeoutId) {
     clearTimeout(this.timeoutId);
     this.timeoutId = null;
   }
 
-  this.timeoutId = setTimeout(function() {
-    _this.timeoutId = null;
-    if (_this.exited) return;
-    _this.exited = true;
-    _this.sandcastle.kickOverSandCastle();
-    _this.emit('timeout', methodName);
+  this.timeoutId = setTimeout(() => {
+    this.timeoutId = null;
+    if (this.exited) { return; }
+    this.exited = true;
+    this.sandcastle.kickOverSandCastle();
+    this.emit('timeout', methodName);
   }, this.timeout);
 };
 
 Script.prototype.reset = function() {
-  if (this.timeoutId){
+  if (this.timeoutId) {
     clearTimeout(this.timeoutId);
     this.timeoutId = null;
   }
   this.exited = false;
 };
 
 Script.prototype.createClient = function(methodName, globals) {
-
-  var _this = this;
-
   try {
-    globals = EJSON.stringify(globals)
-  } catch(err) {
-    _this.onExit(methodName, EJSON.stringify({
+    globals = EJSON.stringify(globals);
+  } catch (err) {
+    this.onExit(methodName, EJSON.stringify({
       error: { message: 'Unserializable Payload' }
     }));
     return;
   }
+  this.sandcastle.sandboxReady(() => {
+    if (this.exited) { return; }
 
-  this.sandcastle.sandboxReady(function() {
-
-    if (_this.exited) return;
-
-    var stream = new BufferStream({ size: 'flexible' });
-
-    var client = net.createConnection(_this.sandcastle.getSocket(), function() {
-      client.write(EJSON.stringify({
-        source: _this.source,// the untrusted JS.
-        sourceAPI: _this.sourceAPI,// the trusted API.
+    const client = net.connect(this.sandcastle.getSocket(), () => {
+      let pack = EJSON.stringify({
+        source: this.source, // the untrusted JS.
+        sourceAPI: this.sourceAPI, // the trusted API.
         globals: globals, // trusted global variables.
         methodName: methodName
-      }) + '\u0000\u0000'); // the chunk separator
+      });
+      // timestamp object in case of collision
+      if (this.sandcastle.sharedObject.isClosed()) {
+        this.sandcastle.createMmap();
+      }
+      // handle resize issue due to roughly 1/3 of free memory
+      if (((Buffer.byteLength(pack) + 400)*3) > this.sandcastle.sharedObject.get_free_memory()) {
+        this.onExit(methodName, EJSON.stringify({
+          error: { message: 'Script and Payload exceeds your memory limit.' }
+        }));
+        return;
+      }
+      pack = Buffer.from(pack);
+      const nanoSecs = process.hrtime();
+      this.timeStamp = ((nanoSecs[0]*1e9) + nanoSecs[1]).toString(36);
```

Comment:
you are both setting this timestamp here (which gets called once per execution), and on line 191 (which gets called N times per execution). This means that the timestamp value is potentially going to be altered many times per script, which means that your cleanup code on line 161 won't actually delete all the correct items from the mmap